File: C++Implementation/BinarySearch.cpp
// Binary Search Algorithm in C++. We will be using Merge Sort to sort array prior to performing binary search. 
// Time Complexity is O(log n) and Space Complexity is O(1)
// This algorithm is a fundamental search algorithm every programmer should know.

#include <iostream>
using namespace std;

// Function Declarations
int* mergesort(int[], int);
int* merge(int*, int*, int, int);
int binarySearch(int*, int, int);

int main() {

	int myArray[] = { 24, 67, 50, 8, 29, 116, 125, 48, 110, 82, 145, 6, 63, 59, 80, 26, 105, 33, 15, 129 };
	int arraySize = size(myArray);
	int* sortedArray; // Create a pointer which will hold first address of sortedArray
	sortedArray = mergesort(myArray, arraySize); // mergesort will return pointer representing first element of sorted array
	for (int i = 0; i < arraySize; i++) {
		cout << sortedArray[i] << " ";
	}
	
	// Call binary search algorithm. Returns index if found. If not found, index of -1 is returned.
	int index = binarySearch(sortedArray, 105, arraySize);
	if (index == -1)
		cout << "\nNot Found";
	else
		cout << "\nFound at index " << index;

	return 0;
}

// mergesort takes two arguments: the array i.e. address of first element in array and arraySize
int* mergesort(int myArray[], int arraySize) {

	// Base condition. If the array has a single element, then the recursive function will return the single element in array
	if (arraySize < 2) {
		return myArray;
	}
	else {
		int middle = arraySize / 2; // The middle value will tell us where to split the array from.
		// Determine the size of each sublist
		int leftSize = middle;
		int rightSize = arraySize - middle;
		// Use the sizes to construct the sublists of specified lengths
		int* left, * right;
		left = new int[leftSize];
		right = new int[rightSize];
		// Use a for loop to create the different splits
		for (int i = 0; i < leftSize; i++) {
			left[i] = myArray[i];
		}
		for (int i = 0; i < rightSize; i++) {
			right[i] = myArray[i + middle];
		}
		// Utilize recursion and call mergesort using the split arrays.
		int* newLeft, * newRight, * newList;
		newLeft = mergesort(left, leftSize);
		newRight = mergesort(right, rightSize);
		// Call merge which will merge the two arrays to one large array
		newList = merge(newLeft, newRight, leftSize, rightSize);
		return newList;
	}
}

// Merge takes the address of the first element in the left and right subarray. It also takes their respective sizes.
int* merge(int* left, int* right, int leftSize, int rightSize) {
	int* mergedList;
	mergedList = new int[leftSize + rightSize]; // Create a merged list which is the size of the sum of left Array Size and right Array Size

	int i = 0, j = 0, k = 0;
	// While the indexes are less than their respective array lengths, add new values to merged list.
	while ((i < leftSize) && (j < rightSize)) {
		// We will add the smallest value between two values from the left subarray and right subarray
		if (left[i] <= right[j]) {
			mergedList[k] = left[i];
			i++;
			k++;
		}
		else {
			mergedList[k] = right[j];
			j++;
			k++;
		}
	}
	// The last two while loops are designed to deal with any values that have been left over
	while (i < leftSize) {
		mergedList[k] = left[i];
		i++;
		k++;
	}
	while (j < rightSize) {
		mergedList[k] = right[j];
		j++;
		k++;
	}
	return mergedList;
}

int binarySearch(int* mySortedArray, int value, int size) {
	bool found = false; // if found, will set to true, else remains false
	// first and last are necessary to calculate middle. index will be used to return location where item is found or -1
	int first = 0, last = size, index = -1;
	while ((found == false) && (first <= last)) {
		int middle = (first + last) / 2; // Calculate middle
		if (middle > size - 1) { // Case where middle is greater than length of array
			break;
		}
		if (mySortedArray[middle] == value) { // If exact value is found, set index to middle and found to true
			found = true;
			index = middle;
		}
		else if (value < mySortedArray[middle]) // If value < mySortedArray[middle], everything to the right does not matter
			last = middle - 1;
		else
			first = middle + 1; // If value > mySortedArray[middle], everything to the left does not matter
	}

	return index;
}

File: PythonImplementation/BinarySearch.py
# Binary Search Algorithm is one of the most important and fundamental search algorithms every programmer should know.
# In order to perform binary search, it is crucial that the list is sorted first.
# We will be using our merge sort algorithm to sort our list.
# Then, we will perform binary search.
# Time Complexity is O(log n). Space Complexity is O(1)

def main():
    myList = [83, 50, 14, 39, 59, 127, 75, 43, 34, 125, 27, 18, 23, 4, 28, 60, 36, 61, 123, 95, 37, 103, 19, 116, 99]
    mySortedList = mergesort(myList)
    # Now that we have obtained our sorted list, we can call the binary search function to perform search. 
    # If found, index will be returned. Else, -1 will be returned.
    print(mySortedList)
    index = binarySearch(mySortedList, 23)
    if index == -1:
        print("Not Found")
    else:
        print("Found at index " + str(index))


# In order to perform mergesort, the mergesort function utilizes a concept called recursion. Recursion is when a function calls itself and is useful for situations such as this.
def mergesort(a_list):
    # In recursive functions, there is a base condition which determines when function should exit recursion
    # In our case, when the len of the list is less than 2, return the list
    if len(a_list) < 2:
        return a_list
    else:
        middle = int(len(a_list) / 2) # The middle represents a point where we split the list into two sublists
        left = a_list[:middle] # Create the left sublist by splitting on middle index
        right = a_list[middle:] # Create the right sublist by splitting on midde index
        newLeft = mergesort(left) # Recursive call sending left sublist
        newRight = mergesort(right) # Recursive call sending right sublist
        newList = merge(newLeft,newRight) # Merge will merge the two sorted sublists
        return newList

def merge(list_a, list_b):
    # These variables will represent the indexes used to merge the two sublists to one whole list
    i = 0
    j = 0
    mergedList = []
    # As long as the length of the list has not been met, we will compare each value and decide which value to put before
    while i < len(list_a) and j < len(list_b):
        if list_a[i] <= list_b[j]: # If this condition is met, list_a[i] will be put into mergedList and index will be incremented.
            mergedList.append(list_a[i])
            i += 1
        else:
            mergedList.append(list_b[j]) # If condition was not met, list_b[i] will be put into mergedList and index will be incremented
            j += 1
    # This next portion will handle any thing that was left over in list_a or list_b
    # Since the previous while loop exited due to one of the conditions failing, only one of the while loops on the bottom will run.
    while i < len(list_a):
        mergedList.append(list_a[i])
        i += 1
    while j < len(list_b):
        mergedList.append(list_b[j])
        j += 1
    return mergedList


def binarySearch(mySortedList, value):
    found = False # If found, will become true
    index = -1 # If never found, index returned will be -1
    # First and Last are necessary to calculate middle point
    first = 0 
    last = len(mySortedList)
    # If the value has not been found and first <= last, check to see if value is in list
    while not found and first <= last:
        middle = int((first + last)/2) # We will use the middle value to find the value
        # It is possible for middle to end up being greater than the length - 1. If thats the case, the value was not found
        if(middle > len(mySortedList) - 1):
            break
        if mySortedList[middle] == value: # If middle index is where value is located, set found to True and set index to middle
            found = True
            index = middle
        elif mySortedList[middle] < value: # If value in middle index is less than value, everything to left of middle does not matter
            first = middle + 1
        else:
            last = middle - 1 # If value in middle index is greater than value, everything to right of middle does not matter.
    return index


main()
